std/internal/mmm.d

```diff
@@ -0,0 +1,337 @@
+/**
+This module provides OOM-checked and memory safe interfaces to C's $(D malloc)
+and $(D calloc) functions.
+
+Acronyms:
+$(B OOM:) $(I Out Of Memory), the condition in which an allocator has no more
+memory space to offer.
+
+Copyright: Jakob Øvrum 2015
+
+License: $(HTTP boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Jakob Øvrum
+*/
+module std.internal.mmm;
+
+import core.stdc.stdlib : calloc, malloc;
+import std.traits : hasElaborateDestructor, hasIndirections;
+import std.range.primitives;
+
+private void[] checkedMallocArray(size_t n) nothrow @system @nogc
+{
+    import core.exception : onOutOfMemoryError;
+
+    if (auto p = malloc(n))
+        return p[0 .. n];
+    else
+    {
+        onOutOfMemoryError();
+        assert(false);
+    }
+}
+
+/**
+ * Allocate space for one or more objects on the C heap with $(D malloc).
+ *
+ * The new objects are uninitialized;
+ * initialize them with $(XREF conv, emplace) before use.
+ * $(D @safe) only when T has no indirections.
+ *
+ * Params:
+ *    T = type of object(s)
+ *    n = number of objects to allocate space for
+ * Throws:
+ *   $(D OutOfMemoryError) on OOM
+ */
+T* checkedMalloc(T)() @system
+    if (hasIndirections!T)
+{
+    return cast(T*)checkedMallocArray(T.sizeof).ptr;
+}
+
+/// Ditto
+T* checkedMalloc(T)() @trusted
+    if (!hasIndirections!T)
+{
+    return cast(T*)checkedMallocArray(T.sizeof).ptr;
+}
+
+/// Ditto
+T[] checkedMallocArray(T)(size_t n) @system
+    if (hasIndirections!T)
+{
+    return cast(T[])checkedMallocArray(T.sizeof * n);
+}
+
+/// Ditto
+T[] checkedMallocArray(T)(size_t n) @trusted
+    if (!hasIndirections!T)
+{
+    return cast(T[])checkedMallocArray(T.sizeof * n);
+}
+
+/**
+ * Construct a new T on the C heap with newly allocated space from $(D malloc).
+ *
+ * Params:
+ *    T = type to construct
+ *    args = construction arguments for T
+ * Throws:
+ *    $(D OutOfMemoryError) on OOM
+ */
+T* checkedMalloc(T, Args...)(auto ref Args args) // safety inferred
```

Comment:
So this would be like `mallocConstruct` because it needs to be differentiated from the other one (e.g. if one wants to have a default-initialized type).